DirectProgramming/DPC++/N-BodyMethods/Nbody/src/GSimulation.cpp

```diff
@@ -0,0 +1,234 @@
+  // Create SYCL buffer for the ener array
+  buffer ebuf(energy, r, {cl::sycl::property::buffer::use_host_ptr()});
+
+  TimeInterval t0;
+  int nsteps = get_nsteps();
+  // Looping across integration steps
+  for (int s = 1; s <= nsteps; ++s) {
+    TimeInterval ts0;
+    // Submitting first kernel to device which computes acceleration of all particles
+    q.submit([&](handler& h) {
+       auto p = pbuf.get_access<access::mode::read_write>(h);
+       h.parallel_for(r, [=](id<1> i) {
+         RealType acc0 = p[i].acc[0];
+         RealType acc1 = p[i].acc[1];
+         RealType acc2 = p[i].acc[2];
+         for (int j = 0; j < n; j++) {
+           RealType dx, dy, dz;
+           RealType distance_sqr = 0.0f;
+           RealType distance_inv = 0.0f;
+
+           dx = p[j].pos[0] - p[i].pos[0];  // 1flop
+           dy = p[j].pos[1] - p[i].pos[1];  // 1flop
+           dz = p[j].pos[2] - p[i].pos[2];  // 1flop
+
+           distance_sqr =
+               dx * dx + dy * dy + dz * dz + kSofteningSquared;  // 6flops
+           distance_inv = 1.0f / sycl::sqrt(distance_sqr);         // 1div+1sqrt
+
+           acc0 += dx * kG * p[j].mass * distance_inv * distance_inv *
+                   distance_inv;  // 6flops
+           acc1 += dy * kG * p[j].mass * distance_inv * distance_inv *
+                   distance_inv;  // 6flops
+           acc2 += dz * kG * p[j].mass * distance_inv * distance_inv *
+                   distance_inv;  // 6flops
+         }
+         p[i].acc[0] = acc0;
+         p[i].acc[1] = acc1;
+         p[i].acc[2] = acc2;
+       });
+     })
+        .wait_and_throw();
+	// Second kernel updates the velocity and position for all particles
+    q.submit([&](handler& h) {
+       auto p = pbuf.get_access<access::mode::read_write>(h);
+       auto e = ebuf.get_access<access::mode::read_write>(h);
+       h.parallel_for(r, [=](id<1> i) {
+         p[i].vel[0] += p[i].acc[0] * dt;  // 2flops
+         p[i].vel[1] += p[i].acc[1] * dt;  // 2flops
+         p[i].vel[2] += p[i].acc[2] * dt;  // 2flops
+
+         p[i].pos[0] += p[i].vel[0] * dt;  // 2flops
+         p[i].pos[1] += p[i].vel[1] * dt;  // 2flops
+         p[i].pos[2] += p[i].vel[2] * dt;  // 2flops
+
+         p[i].acc[0] = 0.f;
+         p[i].acc[1] = 0.f;
+         p[i].acc[2] = 0.f;
+
+         e[i] = p[i].mass *
+                (p[i].vel[0] * p[i].vel[0] + p[i].vel[1] * p[i].vel[1] +
+                 p[i].vel[2] * p[i].vel[2]);  // 7flops
+       });
+     })
+        .wait_and_throw();
+	// Third kernel accumulates the energy of this Nbody system
+    q.submit([&](handler& h) {
+       auto e = ebuf.get_access<access::mode::read_write>(h);
+       h.single_task([=]() {
+         for (int i = 1; i < n; i++) e[0] += e[i];
+       });
+     })
+        .wait_and_throw();
+    auto a = ebuf.get_access<access::mode::read_write>();
+    kenergy_ = 0.5 * a[0];
+    a[0] = 0;
+
+    double elapsed_seconds = ts0.Elapsed();
```

Comment:
Why not using just `std::chrono`?

Reply:
oneAPI package ships a dpc_common.hpp header which comprises of both the timer utility which in turn uses Chrono and also the SYCL async exception handler. Sample code is using this utility which is available at /opt/intel/oneapi/dev-utlities/latest/include when you install oneAPI Base Kit package.